src/aircraft_cmd.cpp

```diff
@@ -430,6 +430,38 @@ static void CheckIfAircraftNeedsService(Aircraft *v)
 	} else if (v->current_order.IsType(OT_GOTO_DEPOT)) {
 		v->current_order.MakeDummy();
 		SetWindowWidgetDirty(WC_VEHICLE_VIEW, v->index, WID_VV_START_STOP);
+	} else {
+		bool hangar_in_o = false;
+		const Order *o;
+
+		/* Is there at least an airport with a hangar in the orders? */
+		FOR_VEHICLE_ORDERS(v, o) {
+			if (o->IsType(OT_GOTO_STATION)) {
+				const Station *ost = Station::Get(o->GetDestination());
+				if (CanVehicleUseStation(v, ost) && ost->airport.HasHangar()) {
+					hangar_in_o = true;
+					break;
+				}
+			}
+		}
+
+		if (!hangar_in_o || (!CanVehicleUseStation(v, st) && v->state >= TAKEOFF && v->state <= FLYING)) {
+			/* there's no airport with a hangar in the orders, or the aircraft
+			 * can't use the airport, so look for a nearby hangar */
+			const StationID nearest_hangar = FindNearestHangar(v);
+
+			/* v->tile can't be used here, when aircraft is flying v->tile is set to 0 */
+			TileIndex vtile = TileVirtXY(v->x_pos, v->y_pos);
+
+			if (nearest_hangar != INVALID_STATION && ((!CanVehicleUseStation(v, st) && v->state >= TAKEOFF && v->state <= FLYING) || (!st->airport.HasHangar() &&
+					/* is nearest hangar closer than destination? */
+					DistanceSquare(vtile, Station::Get(nearest_hangar)->airport.tile) <= DistanceSquare(vtile, st->airport.tile)))) {
```

Comment:
Why all the CanVehicleUseStation and HasHangar checks here? Isn't it already established that the current station either does not have a hangar, or cannot be used by the vehicle?

Reply:
That is because of the first if condition:
`if (st->airport.HasHangar() && CanVehicleUseStation(v, st))`

I only know that either `st->airport.HasHangar()` or `CanVehicleUseStation(v, st)` failed. `CheckIfAircraftNeedsService` is called daily, and it's part of an attempt to fix helicopters travelling between two heliports without hangars, but I also intend to maintain the original behaviour for the other cases. There's also situations where airports could be demolished in-between, and knowing this function is called every day, I went with more checks to try avoid abnormal behaviour.